File: kata01_codekata_karate_chop/clojure/src/main.clj
(ns main)

(defn chop [x, arr]
  (let [left (atom 0)
        right (atom (count arr))
        slice_length (fn [] (- @right @left))
        found (atom false)]
    (while (and (> (slice_length) 1)
                (not @found))
      (let [center (/ (+ @left @right) 2)]
        (case (compare (nth arr center) x)
          -1 (reset! left center)
          1 (reset! right center)
          0 (do
              (reset! found true)
              (reset! left center)))))
    @left))

(defn chop_recursive [x, arr]
  (if (= (count arr) 1) 0
    (let [center (/ (+ 0 (count arr)) 2)]
      (case (compare (nth arr center) x)
        -1 (+ center (chop_recursive x (subvec arr center)))
        1 (chop_recursive x (subvec arr 0 center))
        0 center))))

(defn main [opts]
  (printf "chop(5, [1, 3, 5, 7]): %s%n" (chop_recursive 5 [1 3 5 7])))

File: kata01_codekata_karate_chop/rust/src/main.rs
fn chop<T>(x: T, slice: &[T]) -> usize where T: Ord {
  let mut left = 0;
  let mut right = slice.len();
  let count = |left, right| right - left;

  while count(left, right) > 1 {
    let center = (left + right) / 2;
    if slice[center] > x {
      right = center
    }
    else if slice[center] < x {
      left = center
    }
    else {
      return center;
    }
  }

  left
}

fn main() {
  println!("chop(5, [1, 3, 5, 7]): {}", chop(5, &[1, 3, 5, 7]));
}
